Add support second, minute, week, year on DATE_ADD and DATE_SUB

SebLevDev opened the pull request:
The DATE_ADD and DATE_SUB built-in function of Doctrine supports only the "hour", "month" and "day" intervals. I don't see the point of not having all the intervals available. These should be "minute", "second" and "year". I didn't check compatibility with other database vendors rather than MySQL but I don't think it should be a problem

Cfr Also Ticket https://github.com/doctrine/doctrine2/issues/3698

Commit:
Add support second, minute, week, year on DATE_ADD and DATE_SUB

Ocramius commented on lib/Doctrine/ORM/Query/AST/Functions/DateAddFunction.php:
```diff
@@ -51,7 +51,11 @@ public function getSql(SqlWalker $sqlWalker)
                     $this->intervalExpression->dispatch($sqlWalker)
                 );
-
+			case 'minute':
```
I spotted some `\t` here ;-)

Ocramius reviewed (changes requested):
Overall, patch looks good, but misses testing (especially integration with other DBs

Legenyes commented:
Can i do something again ?

Ocramius commented:
Looks good to me: so far.

Assigning to @lcobucci for another pair of eyes.

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -301,6 +301,24 @@ public function testDateAdd()
+    /**
+     * @group DDC-2938
+     */
+    public function testDateAddTime()
```
Please add ` : void`

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -301,6 +301,24 @@ public function testDateAdd()
+    {
+        $arg = $this->_em->createQuery("SELECT DATE_ADD(CURRENT_TIMESTAMP(), 10, 'second') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -301,6 +301,24 @@ public function testDateAdd()
+
+        $arg = $this->_em->createQuery("SELECT DATE_ADD(CURRENT_TIMESTAMP(), 10, 'hour') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -301,6 +301,24 @@ public function testDateAdd()
+
+        $arg = $this->_em->createQuery("SELECT DATE_ADD(CURRENT_TIMESTAMP(), 10, 'minute') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -335,6 +353,24 @@ public function testDateSub()
+    {
+        $arg = $this->_em->createQuery("SELECT DATE_SUB(CURRENT_TIMESTAMP(), 10, 'second') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -335,6 +353,24 @@ public function testDateSub()
+
+        $arg = $this->_em->createQuery("SELECT DATE_SUB(CURRENT_TIMESTAMP(), 10, 'minute') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -335,6 +353,24 @@ public function testDateSub()
+
+        $arg = $this->_em->createQuery("SELECT DATE_SUB(CURRENT_TIMESTAMP(), 10, 'hour') AS add FROM Doctrine\Tests\Models\Company\CompanyManager m")
+                ->getArrayResult();
+        $this->assertTrue(strtotime($arg[0]['add']) > 0);
```
Use `self::assert*()` instead

lcobucci replied:
Wouldn't be better to use a more specific assertion?

lcobucci commented on tests/Doctrine/Tests/ORM/Functional/QueryDqlFunctionTest.php:
```diff
@@ -335,6 +353,24 @@ public function testDateSub()
     /**
+     * @group DDC-2938
+     */
+    public function testDateSubTime()
```
Please add ` : void`

lcobucci reviewed (changes requested):
It LGTM, just some nitpicking. Also, aren't we missing tests for year and week add/sub?

Legenyes commented:
@lcobucci I cannot change for` self::assert*()` and `: void `. 
These changes must be the subject of another PR.

lcobucci commented:
@Legenyes you can change for the code you're inserting.

Legenyes commented:
It's worse. 
Mixing two syntax in the same file is not a good practice.
I will not make this change in this PR

What about BC for `: void` ?

Ocramius commented:
This patch would land in 2.6 only, which is 7.1 only anyway

On 4 Oct 2017 12:08, "Sébastien Lévêque" <notifications@github.com> wrote:

> It's worse.
> Mixing two syntax in the same file is not a good practice.
> I will not make this change in this PR
>
> What about BC for : void ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/6742#issuecomment-334110142>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakIsNIeuJfwNJ6M0wMq73_S22-w2Cks5so1kmgaJpZM4Pqbxg>
> .
>

lcobucci commented:
@Legenyes `self::assert*()` has already been fixed for v3 on the entire codebase, this means that new code should adhere to that, otherwise we need to reapply things over and over again. In the other hand, if we apply the change on v2 as well we'll be creating conflicts.

I agree that it's ideal to have mixed syntaxes but the least worse solution given this context.

Legenyes commented:
What' news ?

lcobucci commented:
@Legenyes we still don't have tests for `week` and `year` and we still have the same `self::assertTrue()` checks that IMO are not really helpful since it doesn't provide you enough info about the operation that has been done.

greg0ire commented:
[The docs](http://doctrine-orm.readthedocs.io/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html) currently say:

> DATE_ADD(date, days, unit) - Add the number of days to a given date. (Supported units are DAY, MONTH)

please change that too (add your units, and change "number of days")

greg0ire commented:
Closes #5835 ?

lcobucci committed:
Improve tests for DATE_ADD and DATE_SUB

So that we can do proper assertions and cover all the possibilities
of the functions.

JoppeDC commented:
This wasn't in the docs yet, leading me to a search for how to do it. Luckily i found this old PR. I've created a PR to add this to the docs. 
https://github.com/doctrine/orm/pull/7729
